zproject/backends.py

```diff
@@ -624,6 +628,20 @@ def __init__(self) -> None:
         if settings.DEVELOPMENT and settings.FAKE_LDAP_MODE:  # nocoverage
             init_fakeldap()
 
+        if settings.LDAP_SYNC_GROUPS:
+            realm = Realm.objects.get(id=settings.LDAP_GROUP_SYNC_REALM_ID)
+
+            for group in settings.LDAP_GROUP_SYNC_LIST:
+                if not UserGroup.objects.filter(name=group).exists():
+                    ldap_logger.debug(
+                        "creating zulip group %s from ldap for realm %s", group, realm.name
+                    )
+                    UserGroup.objects.create(
+                        name=group,
+                        description="LDAP import of " + group,
+                        realm=realm,
+                    )
```

Comment:
Also, now I get the point of running this just once per sync rather than repeating it for every user - but doing this in `__init__` doesn't achieve it - we initialize once for each user anyway:
```
def sync_user_from_ldap(user_profile: UserProfile, logger: logging.Logger) -> bool:
    backend = ZulipLDAPUserPopulator()
```

Instead maybe this should  be in a `ensure_user_groups_for_ldap_sync_exist` helper func and called in `zerver/management/commands/sync_ldap_user_data.py` before the loop?
```
@transaction.atomic
def sync_ldap_user_data(
    user_profiles: Collection[UserProfile], deactivation_protection: bool = True
) -> None:
    logger.info("Starting update.")
    try:
        realms = {u.realm.string_id for u in user_profiles}
        ensure_user_groups_for_ldap_sync_exist(realms)  # this will loop over the realms and use `.get_or_create` to make sure groups exist?

        for u in user_profiles:
```

Reply:
I used `settings.LDAP_GROUP_SYNC_REALM_ID` (temporary until per-realm config is implemented), otherwise we have a chicken-egg problem with user not being synced because group creation fails and groups are not created because there are not yet any user.

Reply:
see https://github.com/zulip/zulip/pull/23039/commits/523594bb29ff4912f0517600c7e06a7c9b7f4e53